content/en/blog/_posts/2022-04-19-csi-nodeexpandsecret.md

```diff
@@ -0,0 +1,197 @@
+for people to design CSI drivers that could cover all resize scenarios.
+
+- At times, the CSI driver needs to check the actual size of the backend block storage (or image)
+  before proceeding with a node-level filesystem expand operation. This avoids false positive returns
+  from the backend storage cluster during filesystem expands.
+
+- When a PersistentVolume represents encrypted block storage (for example using LUKS)
+  you need to provide a passphrase in order to expand the device, and also to make it possible
+  to grow the filesystem on that device.
+
+- For various validations at time of node expansion, the CSI driver has to be connected
+  to the backend storage cluster. If the` nodeExpandVolume` request includes a `secretRef`
+  then the CSI driver can make use of the same and connect to the storage cluster to
+  perform the cluster operations.
+
+## How does it work?
+
+To enable this functionality from this version of Kubernetes, Kubernetes Storage SIG have introduced
+a new feature gate called `CSINodeExpandSecret`. Once the feature gate is enabled
+in the cluster, NodeExpandVolume requests can include a `secretRef` field. The NodeExpandVolume request
+is part of CSI; for example, in a request which has been sent from the Kubernetes
+control plane to the CSI driver.
+The admin can specify these secrets as an opaque parameter in storage class as
+they specifies other CSI secrets today.  The secrets mention in the storage class
+has to carry the key name as shown below
+
+```
+csi.storage.k8s.io/node-expand-secret-name: test-secret
+csi.storage.k8s.io/node-expand-secret-namespace: default`
+```
+
+If feature gates are enabled and storage class carries the above secret configuration,
+the subject CSI provisioner receives the credentials from the Secret as part of the NodeExpansion request.
+
+CSI volumes that require secrets for online expansion will have NodeExpandSecretRef
+field set. If not set, the NodeExpandVolume CSI RPC call will be made without a secret.
+
+
+
+## Trying it out
+
+1. Enable the `CSINodeExpandSecret` feature gate (please refer to
+[Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)).
+
+1. Create a Secret, and then a StorageClass that uses that Secret
+
+Here's an example manifest for a Secret that holds credentials:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: test-secret
+  namespace: default
+data:
+stringData:
+  username: admin
+  password: t0p-Secret
+```
+
+Here's an example manifest for a StorageClass that refers to those credentials:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-hostpath-sc
+parameters:
+  csi.storage.k8s.io/node-expand-secret-name: test-secret   # the name of the Secret
+  csi.storage.k8s.io/node-expand-secret-namespace: default  # the namespace that the Secret is in
+provisioner: hostpath.csi.k8s.io
```

Comment:
Does this provisioner actually work with NodeExpandVolume secret references?

(if not, maybe make up a fictional provisioner, eg `blockstorage.cloudprovider.example`?)

Reply:
yeah.. the hostpath output and logs are from real testing. should be fine. :+1:

Reply:
How does the `hostPath` provisioner actually _use_ these secrets? If it ignores them, that's maybe not an ideal example.

Reply:
actually this feature is from the kubernetes csi client to push the secrets to the CSI driver as part of the NodeExpandVolume call.. The hostPath provisioner is a CSI driver who get those as shown in the example.. in that sense, it looks fine to me keep the current version as an example .. @sftim

Reply:
if really needed I can replace the string `hostpath` to `hostpath-example` or something similar.. please let me know wdyt @sftim

Reply:
Could you explain in a bit more detail (to reviewers, not readers) how a cluster admin works out what they put into `test-secret`? If I changed the contents of that Secret to have empty strings as the values, would it still work?

If so, then we're not really illustrating things end-to-end. Yes, the _feature_ from the KEP is implemented, but the reader cares about the end-to-end story. They're less interested in seeing an integration test that proves that the secret reference gets to the CSI driver.

Is `blockstorage.cloudprovider.example` a plausible CSI driver name / reference? If it is, how about using that?

Reply:
@sftim the nodeExpandSecret follow the same convention or rules of other similar secrets ( for ex: node stage, publish, controller expand) in the SC as described here 
https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html.  The secret values will be sent to the CSI driver and if the credentials are correct to connect or perform the requested operation it will work otherwise fail. 

>Is blockstorage.cloudprovider.example a plausible CSI driver name / reference? If it is, how about using that?
Sure, I have changed the provisioner name to above.. it sound fine to have above as a pluasible CSI driver name.. 
ptal.. thanks.